realm/realm-annotations-processor/src/test/resources/io/realm/RealmDefaultModuleMediator.java

```diff
@@ -116,15 +117,25 @@ public void insertToRealm(Realm realm, RealmModel type, Map<RealmModel, Long> ca
     }
 
     @Override
-    public void insertToRealm(Realm realm, RealmModel type, Iterator<? extends RealmModel> objects, Map<RealmModel, Long> cache) {
-        // This cast is correct because obj is either
-        // generated by RealmProxy or the original type extending directly from RealmObject
-        @SuppressWarnings("unchecked") Class<RealmModel> clazz = (Class<RealmModel>) ((type instanceof RealmObjectProxy) ? type.getClass().getSuperclass() : type.getClass());
-
-        if (clazz.equals(AllTypes.class)) {
-            AllTypesRealmProxy.insert(realm, objects, cache);
-        } else {
-            throw getMissingProxyClassException(clazz);
+    public void insertToRealm(Realm realm, Collection<? extends RealmModel> objects) {
+        Iterator<? extends RealmModel> iterator = objects.iterator();
+        RealmModel object = null;
+        Map<RealmModel, Long> cache = new IdentityHashMap<RealmModel, Long>(objects.size());
+        if (iterator.hasNext()) {
+            //  access the first element to figure out the type
+            object = iterator.next();
+            insertOrUpdateToRealm(realm, object, cache);
```

Comment:
Isn't this a problem? It means that if you have primary keys it will update behind the scene instead of crashing because you tried to insert an object with a similar ID.

We had the same discussion for `copyToRealm` and `copyToRealmOrUpdate` and right now that doesn't automatically switch to copyOrUpdate, so IMO this should have the same semantics.

Personally I think we should revisit those for 2.x, but for now I think they should behave the same way.